src/gt4py/cartesian/gtc/dace/expansion/daceir_builder.py

```diff
@@ -318,167 +367,348 @@ def visit_FieldAccess(
+                    horizontal_extent=horizontal_extent,
+                    k_interval=k_interval,
+                    inside_horizontal_region=inside_horizontal_region,
+                    targets=targets,
+                    **kwargs,
+                ),
+            )
 
-        write_memlets = _get_tasklet_inout_memlets(
+        tasklet = self._condition_tasklet(
             node,
-            get_outputs=True,
             global_ctx=global_ctx,
-            grid_subset=iteration_ctx.grid_subset,
+            symbol_collector=symbol_collector,
+            iteration_ctx=iteration_ctx,
+            horizontal_extent=horizontal_extent,
             k_interval=k_interval,
+            targets=targets,
+            **kwargs,
+        )
+        code_block = self.visit(
+            oir.CodeBlock(body=node.body, loc=node.loc, label=f"while_{id(node)}"),
+            global_ctx=global_ctx,
+            symbol_collector=symbol_collector,
+            iteration_ctx=iteration_ctx,
+            horizontal_extent=horizontal_extent,
+            k_interval=k_interval,
+            targets=targets,
+            **kwargs,
+        )
+        targets.clear()
+        return dcir.WhileLoop(condition=tasklet, body=code_block)
+
+    def visit_Cast(self, node: oir.Cast, **kwargs: Any) -> dcir.Cast:
+        return dcir.Cast(dtype=node.dtype, expr=self.visit(node.expr, **kwargs))
+
+    def visit_NativeFuncCall(self, node: oir.NativeFuncCall, **kwargs: Any) -> dcir.NativeFuncCall:
+        return dcir.NativeFuncCall(
+            func=node.func, args=self.visit(node.args, **kwargs), dtype=node.dtype
         )
 
-        dcir_node = dcir.Tasklet(
-            decls=decls, stmts=stmts, read_memlets=read_memlets, write_memlets=write_memlets
+    def visit_TernaryOp(self, node: oir.TernaryOp, **kwargs: Any) -> dcir.TernaryOp:
+        return dcir.TernaryOp(
+            cond=self.visit(node.cond, **kwargs),
+            true_expr=self.visit(node.true_expr, **kwargs),
+            false_expr=self.visit(node.false_expr, **kwargs),
+            dtype=node.dtype,
         )
 
-        for memlet in [*read_memlets, *write_memlets]:
+    def _fix_memlet_array_access(
+        self,
+        *,
+        tasklet: dcir.Tasklet,
+        memlets: list[dcir.Memlet],
+        global_context: DaCeIRBuilder.GlobalContext,
+        symbol_collector: DaCeIRBuilder.SymbolCollector,
+        **kwargs: Any,
+    ) -> None:
+        for memlet in memlets:
             """
             This loop handles the special case of a tasklet performing array access.
             The memlet should pass the full array shape (no tiling) and
             the tasklet expression for array access should use all explicit indexes.
             """
-            array_ndims = len(global_ctx.arrays[memlet.field].shape)
-            field_decl = global_ctx.library_node.field_decls[memlet.field]
+            array_ndims = len(global_context.arrays[memlet.field].shape)
```

Comment:
This is clearly symptomatic of a bigger issue around `memlet` size and `indexing` in general. Not to be changed here, but let it be written in the larger documentation that we should not have to fix memlets - this should be dealt with before.